# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
df = pd.read_csv('Step3_dataset.csv',index_col = 'Unnamed: 0')
df.head()

Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939           2500      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600           1540      0        0.0   
2     Hilltop Ski Area   Alaska         2090            294      0        0.0   
3     Arizona Snowbowl  Arizona        11500           2300      0        0.0   
4  Sunrise Park Resort  Arizona        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           550.0         1  
1               90.0             0.0         1  
2              152.0            30.0         1  
3              122.0             0.0         0  
4              104.0            80.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df.dtypes

Name                  object
state                 object
summit_elev            int64
vertical_drop          int64
trams                  int64
fastEight            float64
fastSixes              int64
fastQuads              int64
quad                   int64
triple                 int64
double                 int64
surface                int64
total_chairs           int64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object

In [4]:
df1 = pd.concat([df,pd.get_dummies(df['state'])], axis=1).drop('state', axis=1)
df1

Name  summit_elev  vertical_drop  trams  \
0                       Alyeska Resort         3939           2500      1   
1                  Eaglecrest Ski Area         2600           1540      0   
2                     Hilltop Ski Area         2090            294      0   
3                     Arizona Snowbowl        11500           2300      0   
4                  Sunrise Park Resort        11100           1800      0   
..                                 ...          ...            ...    ...   
325               Meadowlark Ski Lodge         9500           1000      0   
326          Sleeping Giant Ski Resort         7428            810      0   
327                   Snow King Resort         7808           1571      0   
328  Snowy Range Ski & Recreation Area         9663            990      0   
329                White Pine Ski Area         9500           1100      0   

     fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0.0          0          2     2       0       0  ...             0   
1          0.0          0          0     0       0       4  ...             0   
2          0.0          0          0     0       1       0  ...             0   
3          0.0          1          0     2       2       1  ...             0   
4          0.0          0          1     2       3       1  ...             0   
..         ...        ...        ...   ...     ...     ...  ...           ...   
325        0.0          0          0     0       1       1  ...             0   
326        0.0          0          0     0       1       1  ...             0   
327        0.0          0          0     1       1       1  ...             0   
328        0.0          0          0     0       1       3  ...             0   
329        0.0          0          0     0       2       0  ...             0   

     South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0               0          0     0        0         0           0   
1               0          0     0        0         0           0   
2               0          0     0        0         0           0   
3               0          0     0        0         0           0   
4               0          0     0        0         0           0   
..            ...        ...   ...      ...       ...         ...   
325             0          0     0        0         0           0   
326             0          0     0        0         0           0   
327             0          0     0        0         0           0   
328             0          0     0        0         0           0   
329             0          0     0        0         0           0   

     West Virginia  Wisconsin  Wyoming  
0                0          0        0  
1                0          0        0  
2                0          0        0  
3                0          0        0  
4                0          0        0  
..             ...        ...      ...  
325              0          0        1  
326              0          0        1  
327              0          0        1  
328              0          0        1  
329              0          0        1  

[330 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df1['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25, random_state=1)

In [7]:
print(X.shape, y.shape)

(330, 58) (330,)


Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [8]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error, mean_squared_error, r2_score
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [9]:
#Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [10]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
print(explained_variance_score(y_test,y_pred),mean_absolute_error(y_test,y_pred),r2_score(y_test,y_pred))

0.9427818561042524 4.823344177064809 0.9425369820827088


In [11]:
from math import sqrt
rmse = sqrt(mean_squared_error(y_test, y_pred))
rmse

6.262254188252109

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [12]:
lm.intercept_

14.513524776512412

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(lm.coef_, X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
Tennessee            20.595079
North Carolina       14.010776
West Virginia        10.070417
Maryland              9.500111
Massachusetts         6.694864
Virginia              6.535344
New Hampshire         4.613200
Connecticut           2.985732
fastEight             2.795441
Rhode Island          2.276764
Pennsylvania          1.940286
AdultWeekday          0.805689
total_chairs          0.721536
Maine                 0.521730
triple                0.496713
quad                  0.236965
Utah                  0.189158
daysOpenLastYear      0.023484
NightSkiing_ac        0.008667
yearsOpen             0.003426
Snow Making_ac        0.002541
vertical_drop         0.001574
averageSnowfall       0.001051
summit_elev          -0.000003
SkiableTerrain_ac    -0.001697
Runs                 -0.010878
projectedDaysOpen    -0.011548
TerrainParks         -0.085999
double               -0.155056
New Jersey           -0.168432
LongestRun_mi        -0.295813
Nevada               -0.304070
fastQuads            -0.314949
Wisconsin            -0.317521
fastSixes            -0.353436
Vermont              -0.463963
Illinois             -0.521956
trams                -0.874478
surface              -1.109664
New York             -1.287509
clusters             -1.447205
Michigan             -1.874696
Washington           -2.073749
Colorado             -3.158780
Wyoming              -3.489753
Alaska               -3.503785
Oregon               -3.658979
California           -3.918708
Missouri             -3.929704
Minnesota            -4.234379
Arizona              -4.730885
Iowa                 -4.810537
South Dakota         -4.818446
Idaho                -4.999354
Indiana              -6.616756
New Mexico           -6.668592
Montana              -7.090454
Ohio                 -7.292454

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
X2 = df.drop(['Name','state','AdultWeekend'], axis=1)
X2_train, X2_test, y2_train, y2_test = train_test_split(X2,y, test_size=0.25, random_state=1)
lm2 = linear_model.LinearRegression()
model = lm2.fit(X2_train,y2_train)

In [16]:
y2_pred = model.predict(X2_test)
print(explained_variance_score(y2_test,y2_pred),mean_absolute_error(y2_test,y2_pred),r2_score(y2_test,y2_pred))

0.9241713964368776 5.639050353390269 0.9236300986148854


In [17]:
mean_absolute_error(y2_test,y2_pred)

5.639050353390269

In [20]:
X2_df = pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)
X2_df

Coefficient
fastEight             2.606555
surface               1.360366
clusters              1.327564
trams                 0.931461
total_chairs          0.874149
AdultWeekday          0.826816
fastSixes             0.703678
double                0.495427
quad                  0.325099
triple                0.203371
fastQuads             0.177300
TerrainParks          0.154340
LongestRun_mi         0.136623
daysOpenLastYear      0.041332
projectedDaysOpen     0.027042
NightSkiing_ac        0.009093
averageSnowfall       0.007177
SkiableTerrain_ac     0.003224
yearsOpen             0.002712
Runs                  0.002211
vertical_drop         0.002170
Snow Making_ac        0.001414
summit_elev           0.000247

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [28]:
X2_train.count()

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
228         4000            350      0        0.0          0          0     0   
283         3640           2610      0        0.0          0          0     0   
188          950            300      0        0.0          0          0     0   
70           800            400      0        0.0          0          0     2   
173         1480           1040      1        0.0          0          2     2   
..           ...            ...    ...        ...        ...        ...   ...   
203         3500           1100      0        0.0          0          0     0   
255         1800            935      0        0.0          0          1     3   
72           975            275      0        0.0          0          0     0   
235         8540           3690      0        0.0          0          5     0   
37         11952           1162      0        0.0          0          0     1   

     triple  double  surface  ...  LongestRun_mi  SkiableTerrain_ac  \
228       0       1        1  ...            0.1               50.0   
283       0       6        2  ...            3.0             1000.0   
188       0       1        2  ...            0.2               35.0   
70        3       0        6  ...            1.0              100.0   
173       1       1        3  ...            2.0              167.0   
..      ...     ...      ...  ...            ...                ...   
203       1       1        2  ...            2.0              110.0   
255       0       2        2  ...            1.0              120.0   
72        2       0        2  ...            1.0               35.0   
235       1       1        1  ...            2.0             1415.0   
37        0       4        2  ...            1.0              800.0   

     Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  \
228             0.0         78.000000       66.0            100.0   
283           192.0        136.000000       63.0            280.0   
188            26.0        115.103943       55.0            125.0   
70            100.0         82.000000       39.0             24.0   
173           167.0         90.000000       54.0             65.0   
..              ...               ...        ...              ...   
203            75.0         65.000000       26.0            175.0   
255           120.0        116.000000       28.0             40.0   
72             35.0        100.000000       22.0             40.0   
235             0.0        305.000000       82.0            550.0   
37              0.0        143.000000       80.0            350.0   

     AdultWeekday  projectedDaysOpen  NightSkiing_ac  clusters  
228     39.000000          90.000000             0.0         1  
283     79.000000         135.000000             0.0         1  
188     57.916957         120.053004            26.0         1  
70      52.000000          90.000000           100.0         1  
173     79.990000         100.000000           167.0         1  
..            ...                ...             ...       ...  
203     50.000000          65.000000             0.0         1  
255     69.000000         100.000000           120.0         1  
72      40.000000         100.000000            35.0         1  
235     57.916957         305.000000           300.0         2  
37      89.000000         136.000000             0.0         0  

[247 rows x 23 columns]

In [50]:
X2_coefs = list(X2_df.index.to_numpy())
added_feature = []
no_features = []
r2_scores = []
adj_r2_scores =[]
for col in range(len(X2_coefs)):
    cols = X2_coefs[:col+1]
    X_train, X_test, y_train, y_test = train_test_split(X2[cols],y, test_size=0.25, random_state=1)
    lm_loop = linear_model.LinearRegression()
    model = lm_loop.fit(X_train,y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test,y_pred)
    p = len(cols)
    n = len(X_test)
    adj_r2 = 1-(1-r2)*(n-1)/(n-p-1)
    added_feature.append(X2_coefs[col])
    no_features.append(col+1)
    r2_scores.append(r2)
    adj_r2_scores.append(adj_r2)
r2_df = pd.DataFrame(list(zip(added_feature,no_features,r2_scores,adj_r2_scores)), columns = ['Added Feature','No. of Features','R2 Score','Adjusted R2 score'])
r2_df

Added Feature  No. of Features  R2 Score  Adjusted R2 score
0           fastEight                1 -0.003104          -0.015488
1             surface                2 -0.003776          -0.028870
2            clusters                3  0.073050           0.037849
3               trams                4  0.326133           0.291575
4        total_chairs                5  0.525320           0.494497
5        AdultWeekday                6  0.924315           0.918340
6           fastSixes                7  0.924842           0.917827
7              double                8  0.924705           0.916565
8                quad                9  0.926543           0.917487
9              triple               10  0.924724           0.914269
10          fastQuads               11  0.924724           0.913061
11       TerrainParks               12  0.925230           0.912412
12      LongestRun_mi               13  0.925751           0.911762
13   daysOpenLastYear               14  0.926099           0.910884
14  projectedDaysOpen               15  0.928218           0.912147
15     NightSkiing_ac               16  0.929917           0.912927
16    averageSnowfall               17  0.927885           0.909024
17  SkiableTerrain_ac               18  0.922685           0.900940
18          yearsOpen               19  0.922450           0.899062
19               Runs               20  0.922319           0.897261
20      vertical_drop               21  0.920208           0.892738
21     Snow Making_ac               22  0.918985           0.889279
22        summit_elev               23  0.923630           0.893859

In [52]:
X3 = X2[['fastEight','surface','clusters','trams','total_chairs','AdultWeekday','fastSixes','double','quad']]
X3_train, X3_test, y3_train, y3_test = train_test_split(X3,y, test_size=0.25, random_state=1)
lm3 = linear_model.LinearRegression()
model = lm3.fit(X3_train,y3_train)

In [53]:
y3_pred = model.predict(X3_test)
print(explained_variance_score(y3_test,y3_pred),mean_absolute_error(y3_test,y3_pred),r2_score(y3_test,y3_pred))

0.9267908921144454 5.5177844769076545 0.9265428768955076


In [54]:
mean_absolute_error(y3_test,y3_pred)

5.5177844769076545

In [55]:
pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
fastEight        6.941948
trams            2.115585
clusters         1.493595
surface          0.817349
AdultWeekday     0.817047
quad             0.813341
fastSixes        0.568390
total_chairs     0.547894
double           0.320557

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped/Added|
| --- | --- | --- | --- |
| Model 1. | 0.943 | 4.823 | 'base_elev' |
| Model 2. | 0.924 | 5.639 | 'state','base_elev' |
| Model 3. | 0.920 | 5.800 | Kept:  'fastEight','surface','clusters','trams','total_chairs','AdultWeekday','fastSixes','double','quad'|

Model Selection:

Since we cannot move Big Mountain Resort to a different mountain or location, Model 3 is the logical choice for a model. It might be helpful for the BMR stakeholders to see the variance in lift-ticket prices across states, but that question is not very relevant to the chair-lift issue at hand. Therefore, we could have eliminated base elevation, summit elevation, and state during the EDA phase, as our question was how to modify BMR's practices, not to examine industry-wide trends in pricing.